python/thunder/rdds/series.py

```diff
@@ -636,6 +636,111 @@ def query(self, inds, var='inds', order='F', onebased=True):
+        sorted and no records have been removed by filtering, then the resulting output should be equivalent
+        to what one would get from calling myImages.saveAsBinarySeries().
+
+        The json config file written out by this method takes the value of the 'nvalues' key from the
+        length of this object's .index attribute. If the index has been modified to have a different length
+        than the number of values in a Series record, then the resulting data files will be read back in
+        incorrectly.
+
+        Parameters
+        ----------
+        outputdirname : string path or URI to directory to be created
+            Output files will be written underneath outputdirname. This directory must not yet exist
+            (unless overwrite is True), and must be no more than one level beneath an existing directory.
+            It will be created as a result of this call.
+
+        overwrite : bool
+            If true, outputdirname and all its contents will be deleted and recreated as part
+            of this call.
+        """
+        import cStringIO as StringIO
+        import struct
+        from thunder.rdds.imgblocks.blocks import SimpleBlocks
+        from thunder.rdds.fileio.writers import getParallelWriterForPath
+        from thunder.rdds.fileio.seriesloader import writeSeriesConfig
+
+        def partitionToBinarySeries(kvIter):
+            """Collects all Series records in a partition into a single binary series record.
+            """
+            keypacker = None
+            firstKey = None
+            buf = StringIO.StringIO()
+            for seriesKey, series in kvIter:
+                if keypacker is None:
+                    keypacker = struct.Struct('h'*len(seriesKey))
+                    firstKey = seriesKey
+                # print >> sys.stderr, seriesKey, series, series.tostring().encode('hex')
+                buf.write(keypacker.pack(*seriesKey))
+                buf.write(series.tostring())
+            val = buf.getvalue()
+            buf.close()
+            # we might have an empty partition, in which case firstKey will still be None
+            if firstKey is None:
+                return iter([])
+            else:
+                label = SimpleBlocks.getBinarySeriesNameForKey(firstKey) + ".bin"
+                return iter([(label, val)])
+
+        writer = getParallelWriterForPath(outputdirname)(outputdirname, overwrite=overwrite)
+
+        binseriesrdd = self.rdd.mapPartitions(partitionToBinarySeries)
+
+        binseriesrdd.foreach(writer.writerFcn)
+        # note that we use len(self.index) here for nvalues! we don't really have a great alternative
+        # to this at the moment, short of calling first(). 
+        writeSeriesConfig(outputdirname, len(self.dims), len(self.index), dims=self.dims.count,
+                          keytype='int16', valuetype=self.dtype, overwrite=overwrite)
```

Comment:
A couple quick points: 
- there's a simple JSON package in matlab, so writing Series from Matlab or Numpy is easy (assuming one knows the spec). I've had some users doing exactly this. Ultimately, this is my primary reason to prefer the flat binary; when I last investigated more Hadoop friendly formats it was hard to find one that could be trivially written from within those packages.
- I really like the idea of something like a Data API in the future, basically a set of bundled routines to export data from alternate packages into Thunder's Series format, and load them in. Good to know the Python version of such a thing could read something with heterogenous type. I'm not sure the Matlab version could. But let's leave this as is for now and keep thinking about it.